crates/bevy_ecs/src/schedule/executor/async.rs

```diff
@@ -0,0 +1,441 @@
+        };
+        UnparkOnDropGuard(async_barrier.clone());
+        // this allows us to effectively yield as if pending if the world doesn't exist rn.
+        let _world = our_thing.1.try_lock().ok()?;
+        // SAFETY: this is safe because we ensure no one else has access to the world.
+        Some(func(our_thing.0))
+    }
+}
+
+/// Allows you to access the ECS from any arbitrary async runtime.
+/// Calls will never return immediately and will always start Pending at least once.
+/// Call this with the same `PersistentTask` to persist SystemParams like Local or Changed
+/// Just use `world_id` if you do not mind a new SystemParam being initialized every time.
+pub fn async_access<P, Func, Out>(
+    task_identifier: impl Into<EcsTask<P>>,
+    schedule: impl ScheduleLabel,
+    ecs_access: Func,
+) -> impl Future<Output = Result<Out, RunSystemError>>
+where
+    P: SystemParam + 'static,
+    for<'w, 's> Func: Clone + FnMut(P::Item<'w, 's>) -> Out,
+{
+    let task_identifier = task_identifier.into();
+    PendingEcsCall::<P, Func, Out>(
+        PhantomData::<P>,
+        PhantomData,
+        Some(ecs_access),
+        (task_identifier.1, schedule.intern()),
+        task_identifier.0,
+    )
+}
+
+impl<T> From<WorldId> for EcsTask<T> {
+    fn from(value: WorldId) -> Self {
+        EcsTask::new(value)
+    }
+}
+
+/// An EcsTask can be re-used in order to persist SystemParams like Local, Changed, or Added
+pub struct EcsTask<T>(AsyncTaskId, WorldId, PhantomData<T>);
+
+impl<T> Clone for EcsTask<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T> Copy for EcsTask<T> {}
+impl<T> EcsTask<T> {
+    /// Generates a new unique PersistentTask that can be re-used in order to persist SystemParams
+    /// like Local, Changed, or Added
+    pub fn new(world_id: WorldId) -> Self {
+        Self(AsyncTaskId::new().unwrap(), world_id, PhantomData)
+    }
+}
+
+struct PendingEcsCall<P: SystemParam + 'static, Func, Out>(
+    PhantomData<P>,
+    PhantomData<Out>,
+    Option<Func>,
+    (WorldId, InternedScheduleLabel),
+    AsyncTaskId,
+);
+
+impl<P: SystemParam + 'static, Func, Out> Unpin for PendingEcsCall<P, Func, Out> {}
+
+impl<P, Func, Out> Future for PendingEcsCall<P, Func, Out>
+where
+    P: SystemParam + 'static,
+    for<'w, 's> Func: FnOnce(P::Item<'w, 's>) -> Out,
+{
+    type Output = Result<Out, RunSystemError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        fn system_state_init<P: SystemParam + 'static>(world: &mut World, task_id: AsyncTaskId) {
+            world.init_resource::<SystemStatePool<P>>();
+            if !world
+                .get_resource::<SystemStatePool<P>>()
+                .unwrap()
+                .0
+                .read()
+                .unwrap()
+                .contains_key(&task_id)
+            {
+                let system_state = SystemState::<P>::new(world);
+                let cq = ConcurrentQueue::bounded(1);
```

Comment:
Do we need concurrent queues here?
We remove and push the SystemState always inside this task, right? so could it just be an Option?

Reply:
Options are not shareable and mutable across threads so it would have to be an arc<mutex<option>> which i think is less clear than a concurrent queue with size 1, thought both would be non-blocking. So if you do think that it is more clear to use an `Arc<Mutex<Option<T>>>` over a `ConcurrentQueue<T>` i am willing to modify the PR.